### PR TITLE
Fix KV create race after delete/purge

### PR DESCRIPTION
**Pull request by fnichol**
This change fixes an issue when using a JetStream K/V store where a user is creating, deleting, and re-creating keys. If the last entry for a key is a `Operation::Delete` or `Operation::Purge`, the initial `self.update()` returns an error, causing the second part of the method to be exercised.

Prior to this change, if the entry was deleted or purged a `kv.put()` call is used which ignores the revision of that last entry. A single writer to the K/V store would succeed (as no other writers would write first) so no problem. However, if 2 writers attempt to create a key, then a second writer *could* call the `kv.put()` before the first writer calls `kv.put()`. This means that *both* writers get an `Ok(revision)` and can assume that they won the creation of the key.

When using a "distributed lock" pattern (that is many writers race to create a key and the first successful writer wins), this above scenario results in potentially more than one writer who believes they have uniquely acquired the distributed lock.

This change replaces the `kv.put()` call to a `kv.update()` call and provides the `revision` from the deleted/purged entry to ensure that no other writer has beaten the caller to this update. This change closes the race period between concurrent writers to between the first update and the second update call with some optimistic write concurrency to detect another writer.

It appears as though this strategy is in effect in the Go client code [kv.Create] implementation.

[kv.Create]: https://github.com/nats-io/nats.go/blob/278f9f188bca4d7bdee283a0e98ab66b82530c60/jetstream/kv.go#L944-L963

### CODE REVIEW

File: async-nats/src/jetstream/kv/mod.rs [hunk 1]
```diff
@@ -1250,6 +1251,7 @@ impl From<UpdateError> for CreateError {
         match error.kind() {
             UpdateErrorKind::InvalidKey => Error::from(CreateErrorKind::InvalidKey),
             UpdateErrorKind::TimedOut => Error::from(CreateErrorKind::Publish),
+            UpdateErrorKind::WrongLastSequence => Error::from(CreateErrorKind::AlreadyExists),
```
Comment (Jarema) [inline async-nats/src/jetstream/kv/mod.rs:1254]: We should not expose Stream internals here, so for consistency, I would rename it.

Maybe `WrongLastRevision`?

**Review by Jarema — CHANGES_REQUESTED** [review]
Hey!

Thank you for catching that.
This looks good. Just one request for naming consistency change.

File: async-nats/src/jetstream/kv/mod.rs (hunk 1 above)
Comment (fnichol) [reply async-nats/src/jetstream/kv/mod.rs]: Ah yes, this makes a lot of sense (the sequence number is the underlying impl for the revision but that should be opaque to the K/V user), I've updated PR commit with a rename of the variant to `WrongLastRevision`.

We were trying to game out in this scenario whether the caller of `kv.create()` would care to know if this situation occurred, but it seemed like simply knowing that the caller didn't successfully create the key themselves (and that another writer may have slipped in an created the key) was enough.

Thanks again for such a quick review!

File: async-nats/src/jetstream/kv/mod.rs [hunk 2]
```diff
@@ -1370,14 +1373,25 @@ impl Display for UpdateErrorKind {
         match self {
             Self::InvalidKey => write!(f, "key cannot be empty or start/end with `.`"),
             Self::TimedOut => write!(f, "timed out"),
+            Self::WrongLastRevision => write!(f, "wrong last sequence"),
```
Comment (Jarema) [inline async-nats/src/jetstream/kv/mod.rs:1376]: You missed the log message change 🙂

**Review by Jarema — CHANGES_REQUESTED** [review]

File: async-nats/src/jetstream/kv/mod.rs (hunk 2 above)
Comment (fnichol) [reply async-nats/src/jetstream/kv/mod.rs:1376]: You are very correct, one moment 😄 

I had a little too much going on this morning and should have re-read the diff

Comment (fnichol) [reply async-nats/src/jetstream/kv/mod.rs]: There we are

**Review by Jarema — APPROVED** [review]
LGTM!

Thank you for your contribution!

**Pull request merged**